File: cli/wizard/i18n.js
/**
 * cli/wizard/i18n.js — Node-side i18n helper
 *
 * Reads shared/locales/<lang>.json — same catalog used by bash i18n.sh.
 * Resolves $JHT_LANG with cascade: env → host.env → default 'en'.
 * Missing key: returns the key (visible in dev, doesn't break UI).
 *
 * Designed for CLI wizards / Node scripts that need the same strings
 * as the bash launcher (welcome messages, errors, common prompts).
 */
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Path strategy: cli/wizard/i18n.js → ../../shared/locales/
// In container: /app/cli/wizard/i18n.js → /app/shared/locales/
const LOCALES_DIR = path.resolve(__dirname, '..', '..', 'shared', 'locales');

let _cache = null;
let _cacheLang = null;

function readKvFile(filePath, key) {
  try {
    const text = fs.readFileSync(filePath, 'utf8');
    for (const line of text.split('\n')) {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith('#')) continue;
      const eq = trimmed.indexOf('=');
      if (eq < 0) continue;
      if (trimmed.slice(0, eq) !== key) continue;
      return trimmed.slice(eq + 1).replace(/^["']|["']$/g, '');
    }
  } catch { /* ENOENT */ }
  return null;
}

function resolveLang() {
  if (process.env.JHT_LANG && process.env.JHT_LANG.trim()) {
    return process.env.JHT_LANG.trim();
  }
  const home = process.env.JHT_HOME || '/jht_home';
  const candidates = [
    process.env.JHT_HOST_ENV_FILE,
    path.join(home, 'host.env'),
    path.join(process.env.HOME || '/root', '.jht', 'host.env'),
  ].filter(Boolean);
  for (const p of candidates) {
    const v = readKvFile(p, 'JHT_LANG');
    if (v) return v;
  }
  return 'en';
}

function loadCatalog(lang) {
  if (_cache && _cacheLang === lang) return _cache;
  const target = path.join(LOCALES_DIR, `${lang}.json`);
  const fallback = path.join(LOCALES_DIR, 'en.json');
  let merged = {};
  try {
    merged = JSON.parse(fs.readFileSync(fallback, 'utf8'));
  } catch { /* missing fallback is bad but recoverable */ }
  if (lang !== 'en' && fs.existsSync(target)) {
    try {
      const overlay = JSON.parse(fs.readFileSync(target, 'utf8'));
      merged = { ...merged, ...overlay };
    } catch { /* malformed JSON, stick to en */ }
  }
  _cache = merged;
  _cacheLang = lang;
  return merged;
}

// Phase 1 of the launch tutorial is English-only. Existing hosts may still
// carry JHT_LANG=it; keep that choice for later copy, but source the approved
// T0-T1 onboarding keys from the English catalog.
function phase1English(key) {
  return /^(host_setup\.|welcome\.|wizard\.welcome_subtitle$|wizard\.aborted_prereqs$|wizard\.cloud\.|wizard\.step\.|wizard\.profile\.intro$|wizard\.provider\.|wizard\.oauth\.|wizard\.team\.(starting|start_failed)$|wizard\.cli\.|wizard\.(outro_done|outro_aborted|start\.ready|start\.done)$)/.test(key);
}

/**
 * Lookup a translated string by key. Falls back to en, then to the key itself.
 * @param {string} key — e.g. 'wizard.welcome_title'
 * @returns {string}
 */
export function t(key) {
  const lang = resolveLang();
  const cat = phase1English(key) ? loadCatalog('en') : loadCatalog(lang);
  const v = cat[key];
  if (typeof v === 'string') return v;
  return key;
}

/**
 * Lookup with printf-style %s substitution.
 * @param {string} key
 * @param  {...any} args
 * @returns {string}
 */
export function tf(key, ...args) {
  let template = t(key);
  for (const a of args) {
    template = template.replace(/%[ds]/, String(a));
  }
  return template;
}

/** Current resolved language (useful for downstream decisions). */
export function currentLang() {
  return resolveLang();
}

File: shared/i18n.py
"""shared/i18n.py — Python i18n helper

Reads shared/locales/<lang>.json — same catalog used by bash i18n.sh and
cli/wizard/i18n.js. Resolves $JHT_LANG with cascade: env → host.env → 'en'.
Missing key: returns the key (visible in dev, doesn't break UI).

Designed for Python scripts (skills, bridges) that need the same strings
as the bash launcher and Node CLI.

Usage:
    from i18n import t, tf
    print(t('welcome.capitano'))
    print(tf('host_setup.swap_low_ram', 4))
"""
from __future__ import annotations

import json
import os
from functools import lru_cache
from pathlib import Path
from typing import Optional


_LOCALES_DIR = Path(__file__).resolve().parent / "locales"


def _read_kv_file(path: Path, key: str) -> Optional[str]:
    try:
        with path.open(encoding="utf-8") as f:
            for line in f:
                stripped = line.strip()
                if not stripped or stripped.startswith("#"):
                    continue
                if "=" not in stripped:
                    continue
                k, _, v = stripped.partition("=")
                if k.strip() == key:
                    return v.strip().strip('"').strip("'")
    except (OSError, UnicodeDecodeError):
        pass
    return None


@lru_cache(maxsize=1)
def _resolve_lang() -> str:
    env_lang = (os.environ.get("JHT_LANG") or "").strip()
    if env_lang:
        return env_lang
    candidates = [
        os.environ.get("JHT_HOST_ENV_FILE"),
        os.path.join(os.environ.get("JHT_HOME", "/jht_home"), "host.env"),
        os.path.join(os.environ.get("HOME", "/root"), ".jht", "host.env"),
    ]
    for raw in candidates:
        if not raw:
            continue
        v = _read_kv_file(Path(raw), "JHT_LANG")
        if v:
            return v
    return "en"


@lru_cache(maxsize=4)
def _load_catalog(lang: str) -> dict:
    target = _LOCALES_DIR / f"{lang}.json"
    fallback = _LOCALES_DIR / "en.json"
    merged: dict = {}
    if fallback.exists():
        try:
            merged = json.loads(fallback.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, OSError):
            merged = {}
    if lang != "en" and target.exists():
        try:
            overlay = json.loads(target.read_text(encoding="utf-8"))
            merged = {**merged, **overlay}
        except (json.JSONDecodeError, OSError):
            pass
    return merged


def t(key: str) -> str:
    """Lookup a translated string by key. Falls back to en, then to the key."""
    lang = _resolve_lang()
    # T0-T1 launch copy is English-only in phase 1. Do not let an old
    # JHT_LANG=it in host.env leak Italian into the recorded onboarding path.
    if (
        key.startswith(("host_setup.", "welcome.", "wizard.checks.", "wizard.cloud.",
                        "wizard.step.", "wizard.provider.", "wizard.oauth.", "wizard.cli."))
        or key in {
            "wizard.welcome_subtitle", "wizard.aborted_prereqs",
            "wizard.profile.intro", "wizard.team.starting",
            "wizard.team.start_failed", "wizard.outro_done",
            "wizard.outro_aborted", "wizard.start.ready", "wizard.start.done",
        }
    ):
        lang = "en"
    cat = _load_catalog(lang)
    v = cat.get(key)
    if isinstance(v, str):
        return v
    return key


def tf(key: str, *args) -> str:
    """Lookup with printf-style %s/%d substitution."""
    template = t(key)
    try:
        return template % args
    except (TypeError, ValueError):
        return template


def current_lang() -> str:
    """Current resolved language (useful for downstream decisions)."""
    return _resolve_lang()
